src/backend/gl/src/lib.rs

```diff
@@ -736,6 +738,83 @@ pub enum Instance {
     Surface(Surface),
 }
 
+use glutin::platform::unix::RawContextExt;
+impl Instance {
+    
+    pub fn create_surface_from_wayland(
+        &self,
+        display: *mut c_void,
+        surface: *mut c_void,
+    ) -> Surface {
+        log::trace!("Creating GL surface from wayland");
+        let context = unsafe {
+            glutin::ContextBuilder::new()
```

Comment:
the way it would ideally work is: we create a "pure" GL context that is not associated to any surface, for `Instance`. Then physical devices and logical devices, queues, are all just references to this main context.

When a surface is created (here), we'd need to create a GL context per surface, but in a way that shares all resources with the main context. https://docs.rs/glium/0.19.0/glium/glutin/struct.ContextBuilder.html#method.with_shared_lists

This scheme assumes that we can share resources and drivers will not freak out, and that we can create a pure context for the `Instance`.

Reply:
OK, so, if I understand correctly, we want to create a global context once, that we will then pass into future context creations with `with_shared_lists(global_context)`. That way every time you create a surface, it shares the resources with the global context.

So could we do that with something like ( in concept ):

```rust
lazy_static! {
    static ref global_context: Arc<Mutex<Option<Context>>> = Arc::new(Mutex::new(None);
}

//...

fn create_surface() {
    let global_context = global_context.lock().unwrap()
    let shared_ctx = if let Some(context) = global_context {
        context
    } else {
        *global_context = Some(glutin::ContextBuilder::new()
            .with_vsync(true));
    };

    glutin::ContextBuilder::new().with_shared_lists(shared_ctx);
}
```

Reply:
Yes, the global context would be a part of the `Instance`, it doesn't need to be in a static :)

Reply:
Most of the functions to create a context require an event loop. The instance could create an event loop when it is created, but there is a limitation of the wayland backend that requires that all shared contexts use the same event loop. That would be a problem because the event loop is a part of the window that is passed in to `create_surface()` and isn't available when we create the instance.

Creating a context with [`build_osmesa()`](https://docs.rs/glutin/0.23.0/glutin/platform/unix/trait.HeadlessContextExt.html#tymethod.build_osmesa) doesn't require an event loop, but I don't know if you would still have the same problem on wayland.

Reply:
The *idea*, innocent in its purity and naivety, is that the root GL context (associated with the Instance) is a *pure* GL context, meaning that it's not associated to any window, and therefore it shouldn't need any events whatsoever.
When I was looking into it, I ended up prototyping support for pure GL contexts in https://github.com/rust-windowing/glutin/pull/942 . It wasn't merged, and I don't know how the situation really advanced since then. Figuring this out, I think, is one of the core puzzle pieces in the whole GL support story, although there might be other approaches too.

`build_osmesa()` isn't helpful for HW acceleration - it's a software context, it can't share resource with real ones.

One (temporary?) workaround could be to have a GL-specific way of creating an `Instance`, where the user would provide the event loop. Yes, it's an easy to miss invariant, but it would work for most cases, and could be "good enough". That would make the instance creation diverging from the other backends, but at least everything else will be the same,

Reply:
> build_osmesa() isn't helpful for HW acceleration - it's a software context, it can't share resource with real ones.

Got it, that makes sense. I always wondered what mesa was. Now I think I get it: a software implementation of of GL.

> One (temporary?) workaround could be to have a GL-specific way of creating an Instance, where the user would provide the event loop.

If we lazily initialized the global context by waiting until the first call to `create_surface()` where we *would* have an event loop from the passed in window, we wouldn't even have to change the API. I think I'll start with that and leave a doc comment explaining it and that we might want to change it.

Reply:
I don't think going lazy is an option here: we'll be asked to enumerate adapters, provide limits and supported extensions, create devices and resources, etc, all before creating a surface. Heck, there are applications that don't even need a surface.

Reply:
OK, I see it now. I missed the fact that `Headless` and `Surface` had a context in them. I'll look around a little bit for a way to make a "Pure" context and if I can't come up with anything on that front I'll try a temporary alternative way to create a GL instance that requires an event loop.

Reply:
I think I see now where `surfman` could be used. I just found the examples on docs.rs, which don't seem to be on the master branch anymore.

It looks like it might actually be really simple to use `surfman` to create the global context, without requiring a window handle. I'm not sure if it does some magic to handle the context sharing on Wayland or not, but its a start, and we could ask the `surfman` maintainer about that.

I'll try it out and see if that works.

Reply:
Thank you for looking into this! There is a lot of context here, and grasping the situation is the main factor of why the problem is still unsolved. You seem to understand the full picture now.